src/Branch.php

```diff
@@ -1,80 +1,66 @@
 
 
+use Salamek\Zasilkovna\Entity\Branch as BranchEntity;
 use Salamek\Zasilkovna\Model\IBranchStorage;
 
-class Branch
+final class Branch
 {
-    /** @var string */
-    private $jsonEndpoint = 'https://www.zasilkovna.cz/api/v3/%s/branch.json';
-
-    /** @var string */
-    private $apiKey;
-
-    /** @var IBranchStorage */
-    private $branchStorage;
-
-    /**
-     * Branch constructor.
-     * @param $apiKey
-     * @param IBranchStorage $branchStorage
-     */
-    public function __construct($apiKey, IBranchStorage $branchStorage)
-    {
-        $this->apiKey = $apiKey;
-        $this->branchStorage = $branchStorage;
-
-        $this->jsonEndpoint = sprintf($this->jsonEndpoint, $this->apiKey);
-
-        $this->initializeStorage();
-    }
-
-    /**
-     * @throws \Exception
-     */
-    public function initializeStorage($force = false)
-    {
-        if (!$this->branchStorage->isStorageValid() || $force)
-        {
-            $result = file_get_contents($this->jsonEndpoint);
-            if (!$result)
-            {
-                throw new \Exception('Failed to open JSON endpoint');
-            }
-
-            $data = json_decode($result, true);
-            if (!$data || !array_key_exists('data', $data))
-            {
-                throw new \Exception('Failed to decode JSON');
-            }
-            
-            $this->branchStorage->setBranchList($data['data']);
-        }
-    }
-
-    /**
-     * @return mixed
-     * @throws \Exception
-     */
-    public function getBranchList()
-    {
-        return $this->branchStorage->getBranchList();
-    }
-
-    /**
-     * @param $id
-     * @return object
-     */
-    public function find($id)
-    {
-        return $this->branchStorage->find($id);
-    }
+	private IBranchStorage $branchStorage;
+
+	private string $jsonEndpoint;
+
+
+	public function __construct(string $apiKey, IBranchStorage $branchStorage)
+	{
+		if (trim($apiKey) === '') {
+			throw new \RuntimeException('API key can not be empty.');
+		}
+		$this->branchStorage = $branchStorage;
+		$this->jsonEndpoint = 'https://www.zasilkovna.cz/api/v3/' . $apiKey . '/branch.json';
+		$this->initializeStorage();
+	}
+
+
+	public function initializeStorage(bool $force = false): void
+	{
+		if ($force || !$this->branchStorage->isStorageValid()) {
+			if (!($result = file_get_contents($this->jsonEndpoint))) {
+				throw new \RuntimeException('Failed to open JSON endpoint');
+			}
+			if (!($data = \json_decode($result, true)) || !array_key_exists('data', $data)) {
+				throw new \RuntimeException('Failed to decode JSON');
+			}
+
+			$this->branchStorage->setBranchList($data['data']);
+		}
+	}
+
+
+	/**
+	 * @return BranchEntity[]
+	 */
+	public function getBranchList(): array
+	{
+		$return = [];
+		foreach ($this->branchStorage->getBranchList() as $branch) {
+			$return[] = new BranchEntity($branch);
+		}
+
+		return $return;
+	}
+
+
+	public function find(int $id): ?BranchEntity
```

Comment:
I think that return type should be Interface for `BranchEntity` and branch entity should be created in each branch storage separately... so ppl can implement their custom branch storage with a custom implementation of  IBranchEntity...
That way you can implement your own branch storage using Doctrine or so...

Reply:
Ok, I implemented `IBranch` interface.

Specific entity type must be resolved in the `Branch` manager because it calls abstract storage. Because this manager is in the Zasilkovna module, in my opinion it is ok.

Which entity will be instanced can be changed by `setHydrateToEntity()` method.